### PR TITLE
[server][github] fix selection of user for prebuilds

### LINKED ISSUE #8128: Prebuilds stop working for team projects when an owner without repo access joins the team
This issue is a result of investigating #7648 and #8118.

Team-project prebuilds depend on the 1st (returned from DB) team-owner's permissions. If that user does not have the right access to the repo, then the prebuilds stop working. This can start happening if another team-owner is added later, after the project has been working for a while.

`findProjectOwner()` is called during `handlePushEvent()`, and the owner is then used to do `fetchConfig()` which will fail silently if the owner does not have permissions, causing the prebuild not to be triggered. 

Examples of these fetchConfig errors can be found in [traces](https://ui.honeycomb.io/gitpod/datasets/gitpod-production/result/w7GTBB3X9id) and [trace-detail](https://ui.honeycomb.io/gitpod/datasets/gitpod-production/result/w7GTBB3X9id/trace/G2MsTX1zGTG?span=3d1edb0ab64a8070). (internal)

https://github.com/gitpod-io/gitpod/blob/28597c5342fd6f9ca7bcc4bd6859928760df9e45/components/server/ee/src/prebuilds/github-app.ts#L358

### To reproduce
- create a team, add a project with prebuilds, confirm that prebuilds are triggered
- invite a new member to the team who does not have access to the project repo
- set the role of the new member to owner
- observe that prebuilds not being triggered as expected anymore 
- remove the new owner from the team
- observe prebuilds resume

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
Related to https://github.com/gitpod-io/gitpod/pull/8093 this PR fixes a bug in the selection of the user account to ran a prebuild. 

**Before** it would select _any_ of the team owners.
**Now** it should first try with the GitHub App installer first as long as that user is still member of the team. Just as a fallback it might try to pick a team member with a github.com connection, perhaps this account has permissions to access the repository. 

## Related Issue(s)

Fixes #8128

## Release Notes

```release-note
[GitHub] Fix the user account picked for a prebuild.
```

### CODE REVIEW

**Comment by svenefftinge** [comment]
The installations db is based on getting two way connection. One from GH one from gitpod and only if the two entries exist we will add a third 'installed' entry. It seems like with the new project flow we usually only get one of the signals but not the other.

File: components/server/ee/src/prebuilds/github-app.ts
Comment (geropl) [inline components/server/ee/src/prebuilds/github-app.ts:358]: I agree with the logic here. But will double check with Svens comment above...

File: components/server/ee/src/prebuilds/github-app.ts
Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/github-app.ts:169]: good point! checking...

Comment (geropl) [reply components/server/ee/src/prebuilds/github-app.ts:169]: There seem to be a couple of installations per day:
```
SELECT *
	FROM d_b_app_installation AS ai
    WHERE ai.state = 'installed'
    ORDER by ai.creationTime DESC
    LIMIT 200;
```
So I'd expect this to work as expected.

Comment (geropl) [reply components/server/ee/src/prebuilds/github-app.ts:169]: There are a lot of entry in `claimed.platform` state, which mark uncomplete registrations. I don't think that is problem; except that we might want to think hope to increase conversion (out of scope here).

**Comment by jldec** [comment]
Thanks for working on this @AlexTugarev 

My main worry with the implementation is that we are still failing silently when something is not right during the prebuild handler e.g. when fetchConfig throws for some reason. Do you think we could log such errors (in addition to tracing like we do now) and include the project/repo/branch/commit context of the push notification. Logging would allow us to capture metrics also.

For the future, we should also report webhook errors back to users e.g. via the dashboard or commit status.

File: components/server/ee/src/prebuilds/github-app.ts
Comment (svenefftinge) [reply components/server/ee/src/prebuilds/github-app.ts:169]: The current code will happily do prebuilds for those incomplete registrations. With this change all of them will silently no longer see prebuilds.

**Review by jldec — CHANGES_REQUESTED** [review]
I have tested this on a preview environment with the GitHub app installed. 
Unfortunately the behavior described in #8128 is not fixed, in fact it seems worse.
Prebuilds now stop working as soon as another member who does not have repo access joins the team. The member does not have to be promoted to owner.

File: components/server/ee/src/prebuilds/github-app.ts
Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/github-app.ts:169]: @svenefftinge, thanks for revealing this! it's indeed a bigger issue. now that we're not encouraging users to install the Gitpod app from the GitHub marketplace, but rather go through `/new` there is need to adjust the "find app installer" logic.

Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/github-app.ts:169]: I just pushed a change which alters `findInstallation` in a way that can be used as intended in this PR, i.e. it will return the latest recorded installation (with a state != uninstalled) which allows to find the installer of a particular app installation. given that we should be able to fix the wrong user account selection.

For completeness, that still isn't our final destination. We want to select any team {member, owner} with proper access to that repository to be able to ready repository details necessary to create a prebuild.

File: components/gitpod-db/src/typeorm/app-installation-db-impl.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/app-installation-db-impl.ts:42]: :+1:

File: components/gitpod-db/src/typeorm/app-installation-db-impl.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/app-installation-db-impl.ts:53]: nit: I'd love to have some reference here as it's the only piece of code that makes the table schema understandable.
Alternatively, an issue for cleaning it/the whole registration process up would be nice.

**Comment by geropl** [comment]
@AlexTugarev New code LGTM :+1: 

Waiting for @jldec to approve as well.
Comments are nits.

**Comment by jldec** [comment]
I have verified that for projects under a team, adding additional team members or owners who do not have access to the project repo, does not block prebuilds for that repo as reported in #8128.

I see a new problem with prebuilds when the project is NOT under a team, but lives under a personal account. In that case **no prebuilds happen for projects for repos from an org**.  (Personal repos are ok). I verified that the app was installed on the org in question.

File: components/server/ee/src/prebuilds/github-app.ts
Comment (jldec) [inline components/server/ee/src/prebuilds/github-app.ts:356]: This function does not properly handle projects which belong to a user account (not under a team).
In that case the function returns no users, which results in no prebuild being run.

**Review by jldec — CHANGES_REQUESTED** [review]
see comment - doesn't work from projects under user accounts.

File: components/gitpod-db/src/typeorm/app-installation-db-impl.ts
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/app-installation-db-impl.ts:53]: 👍🏻

File: components/server/ee/src/prebuilds/github-app.ts
Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/github-app.ts:356]: Thanks for the thorough testing! Fixed and commented.

**Review by jldec — APPROVED** [review]
lgtm

**Comment by AlexTugarev** [comment]
Just squashed the commits and need a new approval, please!

**Review by jldec — APPROVED** [review]

**Pull request merged**
